Uploader/Program.cs

```diff
@@ -0,0 +1,148 @@
+    /// </summary>
+    public static void Main(string[] args)
+    {
+        if (IsAlreadyRunning())
+        {
+            Console.WriteLine("Another instance is already running.");
+            return;
+        }
+        new Uploader(args).Run();
+    }
+
+    private static bool IsAlreadyRunning()
+    {
+        Process current = Process.GetCurrentProcess();
+        // If this process is not running from the exe file, we allow it
+        // Not sure when this happens, though
+        if (Assembly.GetExecutingAssembly().Location.Replace("/", "\\") != current.MainModule.FileName)
+        {
+            return false;
+        }
+
+        Process[] processes = Process.GetProcessesByName(current.ProcessName);
+        foreach (Process process in processes)
+        {
+            if (process.Id != current.Id)
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private Uploader(string[] args)
+    {
+        traceDirectory = ParseArguments(args);
+        FileSystem fileSystem = new FileSystem();
+        config = ReadConfig(fileSystem);
+        IUpload upload = config.CreateUpload(fileSystem);
+        timerAction = new TimerAction(traceDirectory, config, upload, fileSystem);
+    }
+
+    private static Config ReadConfig(FileSystem fileSystem)
+    {
+        logger.Debug("Reading config from {configFile}", Config.ConfigFilePath);
+
+        Config config;
+        try
+        {
+            config = Config.ReadConfig(fileSystem);
+        }
+        catch (Exception e)
+        {
+            logger.Error(e, "Failed to read config file {configPath}", Config.ConfigFilePath);
+            Environment.Exit(1);
+            return null;
+        }
+
+        List<string> errorMessages = config.Validate();
+        if (errorMessages.Count == 0)
+        {
+            return config;
+        }
+
+        logger.Error("Invalid config file {configPath}: {errorMessages}", Config.ConfigFilePath, String.Join("; ", errorMessages));
+        Environment.Exit(1);
+        return null;
+    }
+
+    /// <summary>
+    /// Parses the command line arguments and returns the trace directory.
+    /// </summary>
+    private string ParseArguments(string[] args)
+    {
+        logger.Debug("Parsing arguments {arguments}", args);
+
+        if (args.Length != 1 || args[0] == "--help")
+        {
+            Console.Error.WriteLine("Usage: Uploader.exe [DIR]");
+            Console.Error.WriteLine("DIR: the directory that contains the trace files to upload.");
+            Console.Error.WriteLine($"The uploader reads its configuration from {Config.ConfigFilePath}");
+            Environment.Exit(1);
+        }
+
+        string traceDirectory = Path.GetFullPath(args[0]);
+        if (!Directory.Exists(traceDirectory))
+        {
+            logger.Error("Directory {traceDirectory} does not exist", traceDirectory);
+            Environment.Exit(1);
```

Comment:
I'm a bit biased by the Env.Exit calls. What you usually do in C# is letting the exception bubble to the main program call (or even beyond) and exit from there.